cpp/docs/man/qpidd.1
.\"
.\" Licensed to the Apache Software Foundation (ASF) under one
.\" or more contributor license agreements.  See the NOTICE file
.\" distributed with this work for additional information
.\" regarding copyright ownership.  The ASF licenses this file
.\" to you under the Apache License, Version 2.0 (the
.\" "License"); you may not use this file except in compliance
.\" with the License.  You may obtain a copy of the License at
.\"
.\"   http://www.apache.org/licenses/LICENSE-2.0
.\"
.\" Unless required by applicable law or agreed to in writing,
.\" software distributed under the License is distributed on an
.\" "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.\" KIND, either express or implied.  See the License for the
.\" specific language governing permissions and limitations
.\" under the License.
.\"

.TH QPIDD "1" "June 2015" "qpidd (qpid-cpp) version 0.34" "User Commands"
.SH NAME

qpidd \- the Qpid AMQP Message Broker Daemon

.SH SYNOPSIS

qpidd [-p port] [--config config_file] [--data-dir directory]

.SH DESCRIPTION

An AMQP message broker daemon that stores, routes and forwards
messages using the Advanced Message Queueing Protocol (AMQP).

.SH OPTIONS

The options below are built-in to qpidd. Installing add-on modules provides additional options. To see the full set of options available type "qpidd --help"

Options may be specified via command line, environment variable or configuration file. See FILES and ENVIRONMENT below for details.

.PP

.SS Options

.TP
\-h [ \-\-help ]
Displays the help message
.TP
\-v [ \-\-version ]
Displays version information
.TP
\-\-config FILE (/etc/qpid/qpidd.conf)
Reads configuration from FILE
\-\-client\-config FILE (/etc/qpid/qpidc.conf) 
Reads client configuration from FILE 
(for cluster interconnect)

.SS Module options
\-\-module\-dir DIR (/usr/lib64/qpid/daemon) 
Load all shareable modules in this 
directory
.TP
\-\-load\-module FILE
Specifies additional module(s) to be 
loaded
.TP
\-\-no\-module\-dir
Don't load modules from module 
directory

.SS Broker Options
\-\-data\-dir DIR (/home/aconway/.qpidd) 
Directory to contain persistent data 
generated by the broker
.TP
\-\-no\-data\-dir
Don't use a data directory.  No 
persistent configuration will be loaded
or stored
.TP
\-\-paging\-dir DIR
Directory in which paging files will be
created for paged queues
.TP
\-p [ \-\-port ] PORT (5672)
Tells the broker to listen on PORT
\-\-interface <interface name>|<interface address> 
Which network interfaces to use to 
listen for incoming connections
\-\-listen\-disable <transport name>     Transports to disable listening
.TP
\-\-worker\-threads N (5)
Sets the broker thread pool size
.TP
\-\-connection\-backlog N (10)
Sets the connection backlog limit for 
the server socket
.TP
\-m [ \-\-mgmt\-enable ] yes|no (1)
Enable Management
.TP
\-\-mgmt\-publish yes|no (1)
Enable Publish of Management Data (\&'no'
implies query\-only)
.TP
\-\-mgmt\-qmf2 yes|no (1)
Enable broadcast of management 
information over QMF v2
.TP
\-\-mgmt\-qmf1 yes|no (0)
Enable broadcast of management 
information over QMF v1
.TP
\-\-mgmt\-pub\-interval SECONDS (10s)
Management Publish Interval
\-\-queue\-purge\-interval SECONDS (600s) 
Interval between attempts to purge any 
expired messages from queues
.TP
\-\-auth yes|no (1)
Enable authentication, if disabled all 
incoming connections will be trusted
.TP
\-\-realm REALM (QPID)
Use the given realm when performing 
authentication
\-\-default\-queue\-limit BYTES (104857600) 
Default maximum size for queues (in 
bytes)
.TP
\-\-tcp\-nodelay
Set TCP_NODELAY on TCP connections
.TP
\-\-require\-encryption
Only accept connections that are 
encrypted
\-\-known\-hosts\-url URL or \&'none' (none) 
URL to send as \&'known\-hosts' to clients
(\&'none' implies empty list)
.TP
\-\-sasl\-config DIR
Allows SASL config path, if supported 
by platform, to be overridden.  For 
default location on Linux, see Cyrus 
SASL documentation.  There is no SASL 
config dir on Windows.
\-\-default\-flow\-stop\-threshold PERCENT (80) 
Percent of queue's maximum capacity at 
which flow control is activated.
\-\-default\-flow\-resume\-threshold PERCENT (70) 
Percent of queue's maximum capacity at 
which flow control is de\-activated.
\-\-default\-event\-threshold\-ratio %age of limit (80) 
The ratio of any specified queue limit 
at which an event will be raised
\-\-default\-message\-group GROUP\-IDENTIFER (qpid.no\-group) 
Group identifier to assign to messages 
delivered to a message group queue that
do not contain an identifier.
.TP
\-\-enable\-timestamp yes|no (0)
Add current time to each received 
message.
\-\-link\-maintenance\-interval SECONDS (2s) 
Interval to check link health and 
.TP
                                      re\-connect
if need be
\-\-link\-heartbeat\-interval SECONDS (120s) 
Heartbeat interval for a federation 
link
\-\-max\-negotiate\-time MILLISECONDS (10000) 
Maximum time a connection can take to 
send the initial protocol negotiation
.TP
\-\-federation\-tag NAME
Override the federation tag

.SS Logging options
.TP
\-t [ \-\-trace ]
Enables all logging
.TP
\-\-log\-enable RULE (notice+)
Enables logging for selected levels and
components. RULE is in the form 
\&'LEVEL[+\-][:PATTERN]'
LEVEL is one of: 
trace debug info notice warning error 
critical
PATTERN is a logging category name, or 
a namespace\-qualified function name or 
name fragment. Logging category names 
are: 
Security Broker Management Protocol 
System HA Messaging Store Network Test 
Client Model Unspecified
For example:
\&'\-\-log\-enable warning+'
logs all warning, error and critical 
messages.
\&'\-\-log\-enable trace+:Broker'
logs all category \&'Broker' messages.
\&'\-\-log\-enable debug:framing'
logs debug messages from all functions 
with \&'framing' in the namespace or 
function name.
This option can be used multiple times
.TP
\-\-log\-disable RULE
Disables logging for selected levels 
and components. RULE is in the form 
\&'LEVEL[+\-][:PATTERN]'
LEVEL is one of: 
trace debug info notice warning error 
critical
PATTERN is a logging category name, or 
a namespace\-qualified function name or 
name fragment. Logging category names 
are: 
Security Broker Management Protocol 
System HA Messaging Store Network Test 
Client Model Unspecified
For example:
\&'\-\-log\-disable warning\-'
disables logging all warning, notice, 
info, debug, and trace messages.
\&'\-\-log\-disable trace:Broker'
disables all category \&'Broker' trace 
messages.
\&'\-\-log\-disable debug\-:qmf::'
disables logging debug and trace 
messages from all functions with 
\&'qmf::' in the namespace.
This option can be used multiple times
.TP
\-\-log\-time yes|no (1)
Include time in log messages
.TP
\-\-log\-level yes|no (1)
Include severity level in log messages
.TP
\-\-log\-source yes|no (0)
Include source file:line in log 
messages
.TP
\-\-log\-thread yes|no (0)
Include thread ID in log messages
.TP
\-\-log\-function yes|no (0)
Include function signature in log 
messages
.TP
\-\-log\-hires\-timestamp yes|no (0)
Use hi\-resolution timestamps in log 
messages
.TP
\-\-log\-category yes|no (1)
Include category in log messages
.TP
\-\-log\-prefix STRING
Prefix to prepend to all log messages

.SS Logging sink options
.TP
\-\-log\-to\-stderr yes|no (1)
Send logging output to stderr
.TP
\-\-log\-to\-stdout yes|no (0)
Send logging output to stdout
.TP
\-\-log\-to\-file FILE
Send log output to FILE.
.TP
\-\-log\-to\-syslog yes|no (0)
Send logging output to syslog;
customize using \-\-syslog\-name and 
\-\-syslog\-facility
.TP
\-\-syslog\-name NAME (qpidd)
Name to use in syslog messages
\-\-syslog\-facility LOG_XXX (LOG_DAEMON) 
Facility to use in syslog messages

.SS Daemon options
.TP
\-d [ \-\-daemon ]
Run as a daemon. Logs to syslog by 
default in this mode.
.TP
\-\-transport TRANSPORT (tcp)
The transport for which to return the 
port
.TP
\-\-pid\-dir DIR (/home/aconway/.qpidd)
Directory where port\-specific PID file 
is stored
.TP
\-w [ \-\-wait ] SECONDS (600)
Sets the maximum wait time to 
initialize or shutdown the daemon. If 
the daemon fails to initialize/shutdown
, prints an error and returns 1
.TP
\-c [ \-\-check ]
Prints the daemon's process ID to 
stdout and returns 0 if the daemon is 
running, otherwise returns 1
.TP
\-q [ \-\-quit ]
Tells the daemon to shut down

.TP
\-\-socket\-fd FD
File descriptor for tcp listening socket

.SS ACL Options
.TP
\-\-acl\-file FILE
The policy file to load from, loaded from 
data dir
.TP
\-\-connection\-limit\-per\-user N (0)
The maximum number of connections allowed 
per user. 0 implies no limit.
.TP
\-\-max\-connections N (500)
The maximum combined number of connections
allowed. 0 implies no limit.
.TP
\-\-connection\-limit\-per\-ip N (0)
The maximum number of connections allowed 
per host IP address. 0 implies no limit.
.TP
\-\-max\-queues\-per\-user N (0)
The maximum number of queues allowed per 
user. 0 implies no limit.

.SS SSL Settings
.TP
\-\-ssl\-use\-export\-policy
Use NSS export policy
.TP
\-\-ssl\-cert\-password\-file PATH
File containing password to use for 
accessing certificate database
.TP
\-\-ssl\-cert\-db PATH
Path to directory containing certificate
database
.TP
\-\-ssl\-cert\-name NAME (gonzo)
Name of the certificate to use
.TP
\-\-ssl\-port PORT (5671)
Port on which to listen for SSL 
connections
.TP
\-\-ssl\-require\-client\-authentication
Forces clients to authenticate in order 
to establish an SSL connection
.TP
\-\-ssl\-sasl\-no\-dict
Disables SASL mechanisms that are 
vulnerable to passive dictionary\-based 
password attacks

.SH FILES
.I /etc/qpidd.conf
.RS
Default configuration file.
.RE

Configuration file settings are over-ridden by command line or environment variable settings. '--config <file>' or 'export QPID_CONFIG=<file>' specifies an alternate file.

Each line is a name=value pair. Blank lines and lines beginning with # are ignored. For example:

  # My qpidd configuration file.
  port=6000
  max-connections=10
  log-to-file=/tmp/qpidd.log

.SH ENVIRONMENT
.I QPID_<option>
.RS
There is an environment variable for each option.
.RE

The environment variable is the option name in uppercase, prefixed with QPID_ and '.' or '-' are replaced with '_'. Environment settings are over-ridden by command line settings. For example:

  export QPID_PORT=6000
  export QPID_MAX_CONNECTIONS=10
  export QPID_LOG_TO_FILE=/tmp/qpidd.log

.SH AUTHOR

The Apache Qpid Project, dev@qpid.apache.org

.SH REPORTING BUGS

Please report bugs to users@qpid.apache.org


cpp/RELEASE_NOTES
Apache Qpid C++ 0.8 Release Notes
--------------------------------

The Qpid 0.8 release of the c++ client and broker support the 0-10
version of the AMQP specification.  You can access this specification
from: 

http://www.amqp.org/confluence/display/AMQP/AMQP+Specification

Please take time to go through the README file provided with the
distro to get a good understanding about build system etc.


cpp/src/qpidd.cpp
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

#include "./qpidd.h"
#include "qpid/Plugin.h"
#include "qpid/Version.h"
#include "qpid/log/Logger.h"
#include "qpid/log/Statement.h"

#include <iostream>
#include <memory>
using namespace std;

namespace qpid {
namespace broker {

auto_ptr<QpiddOptions> options;

// Broker real entry; various system-invoked entrypoints call here.
int run_broker(int argc, char *argv[], bool hidden)
{
    try
    {
        BootstrapOptions bootOptions(argv[0]);
        string           defaultPath (bootOptions.module.loadDir);

        // --version causes print and exit
        if (bootOptions.findArg(argc, argv, "version")) {
            cout << "qpidd (" << qpid::product << ") version "
                << qpid::version << endl;
            return 0;
        }

        // --help sets a flag so that its presence is known despite
        // subsequent parse problems.
        bool helpArgSeen = bootOptions.findArg(argc, argv, "help");

        // Parse only the common, load, and log options to see which
        // modules need to be loaded.  Once the modules are loaded,
        // the command line will be re-parsed with all of the
        // module-supplied options.
        try {
            bootOptions.parse (argc, argv, bootOptions.common.config, true);
            if (hidden)
                bootOptions.log.sinkOptions->detached();
            qpid::log::Logger::instance().configure(bootOptions.log);
        } catch (const std::exception& e) {
            if (helpArgSeen) {
                // provide help even when parsing fails
                 bootOptions.usage();
            }
            // Couldn't configure logging so write the message direct to stderr.
            cerr << endl << "Unexpected error: " << e.what() << endl;
            return 1;
        }

        for (vector<string>::iterator iter = bootOptions.module.load.begin();
             iter != bootOptions.module.load.end();
             iter++)
            qpid::tryShlib (*iter);

        if (!bootOptions.module.noLoad) {
            bool isDefault = defaultPath == bootOptions.module.loadDir;
            qpid::loadModuleDir (bootOptions.module.loadDir, isDefault);
        }

        // Parse options.  In the second pass, do not allow unknown options.
        // All the modules have been added now, so any unknown options
        // should be flagged as errors.
        try {
            options.reset(new QpiddOptions(argv[0]));
            options->parse(argc, argv, options->common.config, false);
        } catch (const std::exception& /*e*/) {
            if (helpArgSeen) {
                 // provide help even when parsing fails
                options->usage();
            }
            throw;
        }

        // Options that just print information.
        if (helpArgSeen) {
            options->usage();
            return 0;
        }

        // Everything else is driven by the platform-specific broker
        // logic.
        QpiddBroker broker;
        return broker.execute(options.get());
    }
    catch(const exception& e) {
        QPID_LOG(critical, "Unexpected error: " << e.what());
    }
    return 1;
}
}}
